### PR TITLE
Be more strict about when inventory units and shipments are allowed to be destroyed.

### PR DESCRIPTION
**Pull request by athal7**
The checkout flow is fairly liberal with trying to destroy and rebuild shipments. This puts a few more checks in place to make sure that it doesn't destroy inventory units and shipments that we care about keeping around.

### CODE REVIEW

**Comment by athal7** [comment]
closing, need to put in a few more checks

**Comment by athal7** [comment]
ready for review now

**Comment by athal7** [comment]
@bonobos/engineering @bonobos/free-running-technologies finally have the specs all worked through here, would enjoy a few more sets of eyes

File: api/app/controllers/spree/api/shipments_controller.rb [hunk 1]
```diff
@@ -75,9 +74,14 @@ def remove
         variant = Spree::Variant.find(params[:variant_id])
         quantity = params[:quantity].to_i
 
-        @shipment.order.contents.remove(variant, quantity, @shipment)
-        @shipment.reload if @shipment.persisted?
-        respond_with(@shipment, default_template: :show)
+        unless @shipment.pending?
```
Comment (magnusvk) [inline api/app/controllers/spree/api/shipments_controller.rb]: I never like unless with an else. Can we unwind this and make it a straight if else?

File: core/app/models/spree/inventory_unit.rb [hunk 2]
```diff
@@ -127,11 +127,15 @@ def current_return_item
         return_items.not_cancelled.first
       end
 
-      # If an inventory unit is associated with return items and really needs
-      # to be deleted then explicitly delete the associated return items first.
-      def ensure_no_return_items
-        if return_items.exists?
-          raise "Inventory units associated with return items should not be deleted."
+      def ensure_can_destroy
+        unless backordered? || on_hand?
```
Comment (magnusvk) [inline core/app/models/spree/inventory_unit.rb]: I would prefer `if !backordered? && !on_hand?`. I don't feel as strongly about this one as I do about the unless / else block above though.

**Comment by magnusvk** [comment]
This looks good to me, though I don't have a ton of context here. :+1:

File: api/app/controllers/spree/api/shipments_controller.rb (hunk 1 above)
Comment (athal7) [inline api/app/controllers/spree/api/shipments_controller.rb]: eek yes unless/else stinks, i did not pay attention to the fact that i did that

File: core/app/models/spree/inventory_unit.rb (hunk 2 above)
Comment (athal7) [inline core/app/models/spree/inventory_unit.rb]: that's fair. i think this way reads a bit more clearly to me, so i'll plan to keep it as is for now unless i hear a second objection.

Comment (kamui) [inline core/app/models/spree/inventory_unit.rb]: I'd agree with @magnusvk here. I find `unless` with `||` to be hard to parse and understand.

Comment (athal7) [inline core/app/models/spree/inventory_unit.rb]: cool, updated

File: core/app/models/spree/order.rb [hunk 3]
```diff
@@ -438,9 +438,11 @@ def ensure_shipping_address
     end
 
     def create_proposed_shipments
-      adjustments.shipping.delete_all
-      shipments.destroy_all
-      self.shipments = Spree::Stock::Coordinator.new(self).shipments
+      if !completed? && shipments.all?(&:pending?)
```
Comment (jhawthorn) [inline core/app/models/spree/order.rb]: I think this is a sensible restriction, but I'd prefer this raised an exception rather than failing (doing nothing) silently.

File: core/app/models/spree/order.rb [hunk 4]
```diff
@@ -438,9 +438,11 @@ def ensure_shipping_address
...
+        adjustments.shipping.delete_all
```
Comment (jhawthorn) [inline core/app/models/spree/order.rb]: This does nothing

```
> puts order.adjustments.shipping.to_sql
SELECT "spree_adjustments".* FROM "spree_adjustments" WHERE "spree_adjustments"."adjustable_id" = 2 AND "spree_adjustments"."adjustable_type" = 'Spree::Order' AND "spree_adjustments"."adjustable_type" = 'Spree::Shipment'  ORDER BY spree_adjustments.created_at ASC
```

**Comment by athal7** [comment]
@jhawthorn updated with your suggestions

**Comment by athal7** [comment]
going to merge this since i've addressed the suggestions and i want to stop destroying this data sooner rather than later
